source/common/logger/ur_sinks.h

```diff
@@ -0,0 +1,66 @@
+// Copyright (C) 2023 Intel Corporation
+// SPDX-License-Identifier: MIT
+
+#include <ostream>
+
+namespace sink
+{
+
+class Sink
+{
+public:
+    Sink() {
+        this->fd = stdout;
+    }
+
+    Sink(std::ostream &os)
+    {
+        if (os.rdbuf() == std::cerr.rdbuf()) {
+            this->fd = stderr;
+        }
+    }
+
+    ~Sink() = default;
+
+    template<typename... Args>
+    void log(const char *level, const char *format, Args&&... args) {
+        const int buf_size = 8192;
+        char buffer[buf_size];
+        int buf_size_left = buf_size;
+        unsigned position = 0;
+
+        appendToBuffer(position, buffer, buf_size_left, "[%s]:", level);
+        appendToBuffer(position, buffer, buf_size_left, format, std::forward<Args>(args)...);
+        appendToBuffer(position, buffer, buf_size_left, "\n", level);
+
+	    int ret = std::fputs(buffer, fd);
+        if (ret < 0) {
+            std::fputs("Call to fputs failed!\n", fd);
+        }
+    }
+
+private:
+    std::FILE *fd;
+
+    template<typename... Args>
+    int printToBuffer(char *buffer, const int buf_size, const char *format, Args&&... args) {
+        int ret = std::snprintf(buffer, buf_size, format, std::forward<Args>(args)...);
```

Comment:
Well, I was thinking to get rid of unsafe C-like format at all.
The problem if you use them now in a new project, it will be here for the next decades in this project... :-(
Using something https://fmt.dev/latest/index.html#safety instead and removing it when you switch to C++20?

Reply:
Provided a simple formatting function to avoid adding a new dependency.